cdap-examples/WikipediaPipeline/src/main/java/co/cask/cdap/examples/wikipedia/TopNMapReduce.java

```diff
@@ -0,0 +1,154 @@
+    Job job = context.getHadoopJob();
+    WorkflowToken workflowToken = context.getWorkflowToken();
+    int topNRank = 10;
+    if (runtimeArguments.containsKey("topn.rank"))  {
+      topNRank = Integer.parseInt(runtimeArguments.get("topn.rank"));
+    }
+    if (workflowToken != null) {
+      workflowToken.put("topn.rank", Value.of(topNRank));
+    }
+    int numReduceTasks = 1;
+    if (runtimeArguments.containsKey("num.reduce.tasks")) {
+      numReduceTasks = Integer.parseInt(runtimeArguments.get("num.reduce.tasks"));
+    }
+    job.setNumReduceTasks(numReduceTasks);
+    job.setMapperClass(TokenizerMapper.class);
+    job.setReducerClass(TopNReducer.class);
+  }
+
+  @Override
+  public void onFinish(boolean succeeded, MapReduceContext context) throws Exception {
+    WorkflowToken workflowToken = context.getWorkflowToken();
+    if (workflowToken != null) {
+      workflowToken.put("result", Value.of(succeeded));
+    }
+  }
+
+  /**
+   * Mapper that emits tokens.
+   */
+  public static class TokenizerMapper extends Mapper<byte [], byte [], Text, IntWritable> {
+    private final Text outputKey = new Text();
+    private final IntWritable one = new IntWritable(1);
+
+    @Override
+    protected void map(byte [] key, byte [] value, Context context) throws IOException, InterruptedException {
+      StringTokenizer tokenizer = new StringTokenizer(Bytes.toString(value));
+      while (tokenizer.hasMoreTokens()) {
+        String word = tokenizer.nextToken().trim();
+        outputKey.set(word);
+        context.write(outputKey, one);
+      }
+    }
+  }
+
+  /**
+   * Reducer that outputs top N tokens. Implements {@link ProgramLifecycle} to demonstrate accessing workflow
+   * token in a Mapper/Reducer class.
+   */
+  public static class TopNReducer extends Reducer<Text, IntWritable, byte [], byte []>
+    implements ProgramLifecycle<MapReduceContext> {
+    private final Map<Text, Integer> countMap = new HashMap<>();
+    private int n = 10;
+
+    @Override
+    public void initialize(MapReduceContext context) throws Exception {
+      WorkflowToken workflowToken = context.getWorkflowToken();
+      if (workflowToken != null) {
+        Value value = workflowToken.get("topn.rank");
+        if (value != null) {
+          n = value.getAsInt();
+        }
+      }
+    }
+
+    @Override
+    protected void reduce(Text key, Iterable<IntWritable> values,
+                          Context context) throws IOException, InterruptedException {
+      int sum = Iterables.size(values);
+      // Need to create a new Text instance because the instance we receive is the same for all records.
+      countMap.put(new Text(key), sum);
+    }
+
+    @Override
+    protected void cleanup(Context context) throws IOException, InterruptedException {
+      List<Map.Entry<Text, Integer>> entries = new ArrayList<>(countMap.entrySet());
+      Collections.sort(entries, new Comparator<Map.Entry<Text, Integer>>() {
```

Comment:
use a TreeMap and pass a comparator. so you don't have to convert to list and do a sort.